src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -2232,15 +2232,47 @@ fn ambiguity_error(
     emit_error: bool,
 ) -> bool {
     let mut descrs = FxHashSet::default();
-    let kinds = candidates
+    // proc macros can exist in multiple namespaces at once,
+    // so we need to compare DefIds to remove
+    // the candidite in the fn namespace
+    let mut possible_proc_macro_id = None;
+    let is_proc_macro_crate = cx.tcx.crate_types() == &[CrateType::ProcMacro];
+    let mut kinds = candidates
         .iter()
-        .map(
-            |(res, def_id)| {
-                if let Some(def_id) = def_id { Res::from_def_id(cx.tcx, *def_id) } else { *res }
-            },
-        )
-        .filter(|res| descrs.insert(res.descr()))
+        .map(|(res, def_id)| {
+            let r =
+                if let Some(def_id) = def_id { Res::from_def_id(cx.tcx, *def_id) } else { *res };
+            if is_proc_macro_crate && let Res::Def(DefKind::Macro(_), id) = r {
```

Comment:
To prevent having this `if` called on each candidate, I tend to prefer creating two functions: one doing nothing and the other having the `let Res::Def` condition. Then based on `is_proc_macro_crate` value, you use one function or the other.

Reply:
Hmm, that would be really tricky here for two reasons:
1. we're not doing nothing in the non proc macro case, we still have the `Res::from_def_id` logic.  This could be gotten around by adding a second `map` call, but:
2. the closure has side effects, so we would need to use a `&mut dyn FnMut`, and I'm pretty sure a vtable dispatch is going to be slower than an if statement (both would need to be done once per candidate).  The if statement should also hopefully play nice with the branch predictor.

Reply:
We can pass the `def_id` and `cx.tcx` as arguments. It can be plain functions, no need for closures.